yt_dlp/extractor/kanal2.py

```diff
@@ -0,0 +1,66 @@
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    join_nonempty,
+    traverse_obj,
+    unified_timestamp,
+    url_or_none,
+)
+
+
+class Kanal2IE(InfoExtractor):
+    _VALID_URL = r'https?://kanal2\.postimees\.ee/[^?]+\?(.*?&)?id=(?P<id>\d+)'
+    _TESTS = [
+        {
+            'note': 'Test standard url (#5575)',
+            'url': 'https://kanal2.postimees.ee/pluss/video/?id=40792',
+            'md5': '7ea7b16266ec1798743777df241883dd',
+            'info_dict': {
+                'id': '40792',
+                'ext': 'mp4',
+                'title': 'Aedniku aabits / Osa 53  (05.08.2016 20:00)',
+                'thumbnail': r're:https?://.*\.jpg$',
+                'description': 'md5:53cabf3c5d73150d594747f727431248',
+                'upload_date': '20160805',
+                'timestamp': 1470427200,
+            }
+        },
+    ]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        playlist = self._download_json(
+            f'https://kanal2.postimees.ee/player/playlist/{video_id}',
+            video_id, query={'type': 'episodes'},
+            headers={'X-Requested-With': 'XMLHttpRequest'})
+
+        return {
+            'id': video_id,
+            'title': join_nonempty(*traverse_obj(playlist, ('info', ('title', 'subtitle'))), delim=' / '),
+            'description': traverse_obj(playlist, ('info', 'description')),
+            'thumbnail': traverse_obj(playlist, ('data', 'image')),
+            'formats': self.get_formats(playlist, video_id),
+            'timestamp': unified_timestamp(self._search_regex(
+                r'\((\d{2}\.\d{2}\.\d{4}\s\d{2}:\d{2})\)$',
+                traverse_obj(playlist, ('info', 'subtitle')), 'timestamp', default=None)),
+        }
+
+    def get_formats(self, playlist, video_id):
+        session = self._download_json(
+            'https://sts.postimees.ee/session/register',
+            video_id, note='Creating session', errnote='Error creating session',
+            headers={
+                'X-Original-URI': traverse_obj(playlist, ('data', 'path')),
```

Comment:
Is this okay to be None?

Reply:
No. http client doesn't like `None` there.

<details open>

```
[Kanal2] 40792: Downloading JSON metadata
[Kanal2] 40792: Creating session
E
======================================================================
ERROR: test_Kanal2 (__main__.TestDownload):
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/glen/scm/yt/yt-dlp/test/test_download.py", line 160, in test_template
    res_dict = ydl.extract_info(
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/YoutubeDL.py", line 1484, in extract_info
    return self.__extract_info(url, self.get_info_extractor(key), download, extra_info, process)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/YoutubeDL.py", line 1495, in wrapper
    return func(self, *args, **kwargs)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/YoutubeDL.py", line 1571, in __extract_info
    ie_result = ie.extract(url)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/common.py", line 678, in extract
    ie_result = self._real_extract(url)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/kanal2.py", line 41, in _real_extract
    'formats': self.get_formats(playlist, video_id),
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/kanal2.py", line 51, in get_formats
    session = self._download_json(
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/common.py", line 1038, in download_content
    res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/common.py", line 1002, in download_handle
    res = self._download_webpage_handle(
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/common.py", line 872, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/extractor/common.py", line 811, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/YoutubeDL.py", line 3719, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 519, in open
    response = self._open(req, data)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 536, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/Users/glen/scm/yt/yt-dlp/yt_dlp/utils.py", line 1527, in https_open
    return self.do_open(
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/urllib/request.py", line 1348, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1323, in _send_request
    self.putheader(hdr, value)
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/http/client.py", line 1259, in putheader
    if _is_illegal_header_value(values[i]):
TypeError: expected string or bytes-like object
```

</details>

Reply:
I tried putting empty string there. it errors out as well. but even worse, spamming site with retries. this could be considered abuse. how could plugin indicate "hey. all is broken. get out"?

the ssl errors are because i ctrl+c

<details open>

```
➜ ./yt-dlp.sh https://kanal2.postimees.ee/pluss/video/?id=40792
[Kanal2] 40792: Downloading JSON metadata
[Kanal2] 40792: Creating session
[Kanal2] 40792: Downloading m3u8 information
[info] 40792: Downloading 1 format(s): 884
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 158
[download] Destination: Aedniku aabits ⧸ Osa 53  (05.08.2016 20_00) [40792].mp4
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (1/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (2/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (3/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (4/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (5/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (6/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (7/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (8/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (9/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 1 (10/10)...
[download] fragment not found; Skipping fragment 1 ...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (1/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (2/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (3/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (4/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (5/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (6/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (7/10)...
Exception ignored in: <ssl.SSLSocket [closed] fd=7, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.8/Frameworks/Python.framework/Versions/3.10/lib/python3.10/ssl.py", line 1234, in sendall
    with memoryview(data) as view, view.cast("B") as byte_view:
ResourceWarning: unclosed <ssl.SSLSocket [closed] fd=7, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (8/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (9/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 2 (10/10)...
[download] fragment not found; Skipping fragment 2 ...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 3 (1/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 3 (2/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 3 (3/10)...
[download] Got error: HTTP Error 403: Forbidden. Retrying fragment 3 (4/10)...
^CException ignored in: <socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.0.36', 50966)>
ResourceWarning: unclosed <socket.socket fd=8, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.0.36', 50966)>

ERROR: Interrupted by user
Exception ignored in: <_io.FileIO name=6 mode='wb' closefd=True>
ResourceWarning: unclosed file <_io.BufferedWriter name=6>
Exception ignored in: <ssl.SSLSocket [closed] fd=7, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
ResourceWarning: unclosed <ssl.SSLSocket [closed] fd=7, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
Exception ignored in: <ssl.SSLSocket [closed] fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
ResourceWarning: unclosed <ssl.SSLSocket [closed] fd=9, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0>
```

</details>

Reply:
see my suggestion below

Reply:
applied